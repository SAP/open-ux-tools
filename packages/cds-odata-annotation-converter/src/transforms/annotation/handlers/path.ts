import type { Path, Identifier, Separator } from '@sap-ux/cds-annotation-parser';
import {
    PATH_TYPE,
    nodeRange,
    copyRange,
    copyPosition,
    IDENTIFIER_TYPE,
    SEPARATOR_TYPE
} from '@sap-ux/cds-annotation-parser';

import type { Element, ODataPathSeparatorDiagnostic } from '@sap-ux/odata-annotation-core-types';
import {
    createElementNode,
    createTextNode,
    Range,
    Edm,
    Diagnostic,
    DiagnosticSeverity,
    ODATA_PATH_SEPARATOR_RULE
} from '@sap-ux/odata-annotation-core-types';

import { i18n } from '../../../i18n';

import type { NodeHandler } from '../handler';
import type { VisitorState } from '../visitor-state';

import { pathLikeTypeElementName } from '../path-utils';

export const pathHandler: NodeHandler<Path> = {
    type: PATH_TYPE,
    convert(state: VisitorState, node: Path): Element | undefined {
        const elementName =
            (state.context.valueType ? pathLikeTypeElementName(state.context.valueType) : undefined) ?? Edm.Path;

        const pathText = node.value
            .split('@')
            .map((segment, i) => (i === 0 ? segment.replace(/\./g, '/') : segment))
            .join('@');

        const fragmentRanges = calculateFragmentRanges(state, node);

        state.addPath(pathText);

        const element: Element = createElementNode({
            name: elementName,
            range: nodeRange(node, true),
            contentRange: nodeRange(node, false),
            content: [createTextNode(pathText, nodeRange(node, false), fragmentRanges)]
        });
        return element;
    }
};

interface Fragment {
    image: string;
    range: Range;
}
/**
 * Fragment ranges are calculated as following.
 * before first @ all '.' chars have been replaced by '/', hence every segment considered as fragment.
 * after first @ fragments are parts separated by '/', i.e. dots are not separators of fragments in that region.
 *
 * @param state - The visitor state.
 * @param node - The path node containing segments and separators.
 * @returns An array of fragment ranges derived from the positions of segments and separators.
 */
function calculateFragmentRanges(state: VisitorState, node: Path): Range[] {
    const tokens = [...node.segments, ...node.separators].sort((a, b) => {
        const startA = a.range?.start?.character;
        const startB = b.range?.start?.character;
        if (startA && startB && startA !== startB) {
            return startA - startB;
        }
        const endA = a.range?.end?.character;
        const endB = b.range?.end?.character;
        if (endA && endB) {
            return endA - endB;
        }
        return 0;
    });

    const fragments = convertToFragments(state, node, tokens);

    return fragments.map((fragment): Range => fragment.range);
}

/**
 * Converts the segments of a Path node into Fragments, handling both data model segments and annotation references.
 *
 * @param state - The VisitorState object managing the state during the visit.
 * @param node - The Path node representing the path.
 * @param tokens - The array of Identifier or Separator tokens representing the path segments.
 * @returns An array of Fragments representing the converted segments of the path.
 */
function convertToFragments(state: VisitorState, node: Path, tokens: (Identifier | Separator)[]): Fragment[] {
    // If any of the token range is not defined,
    // then we can't generate fragment ranges correctly => no ranges are returned
    const fragments: Fragment[] = [];
    let i = 0;

    // Process data model segments (until first annotation reference starting with @)
    for (; i < tokens.length; i++) {
        const token = tokens[i];
        if (token.type === IDENTIFIER_TYPE) {
            if (token.value.includes('@')) {
                // segment referencing annotation is found
                break;
            }
            if (!token.range) {
                return [];
            }

            const fragment: Fragment = {
                image: token.value,
                range: copyRange(token.range)
            };
            fragments.push(fragment);
        } else if (token.value === '/' && !token.escaped) {
            pushWrongPathSeparatorDiagnostic(state, token);
        }
    }

    const annotationFragments = processAnnotationSegments(state, node, tokens, i);
    // TODO: check if we should actually include fragments with 0 length
    // filtering fragments with no content to preserve existing behavior
    return [...fragments, ...annotationFragments].filter((fragment) => fragment.image.length);
}

/**
 * Processes the segments of an annotation path, extracting fragments and checking for missing path escaping.
 *
 * @param state - The VisitorState object managing the state during the visit.
 * @param node - The Path node representing the path.
 * @param tokens - The array of Identifier or Separator tokens representing the path segments.
 * @param startIndex - The index indicating the start of the range in the tokens array.
 * @returns An array of Fragments representing the processed segments of the annotation path.
 */
function processAnnotationSegments(
    state: VisitorState,
    node: Path,
    tokens: (Identifier | Separator)[],
    startIndex: number
): Fragment[] {
    const fragments: Fragment[] = [];
    let requiresEscaping = false;
    let i = startIndex;
    while (i < tokens.length) {
        let token = tokens[i];

        if (!token.range) {
            return [];
        }

        const fragment: Fragment = {
            image: '',
            range: Range.create(copyPosition(token.range.start), copyPosition(token.range.start))
        };

        while (token?.type === IDENTIFIER_TYPE || token?.value === '.') {
            fragment.image += token.value;

            if (fragment.range && token.range) {
                fragment.range.end = copyPosition(token.range.end);
            }

            i++;
            token = tokens[i];

            if (token?.type === SEPARATOR_TYPE && token.value === '/' && !token.escaped) {
                requiresEscaping = true;
            }
        }

        fragments.push(fragment);
        i++;
    }

    if (requiresEscaping) {
        pushMissingPathEscapingDiagnostic(state, node, tokens, startIndex);
    }

    return fragments;
}

/**
 * Pushes a diagnostic for missing path escaping brackets in the specified range of tokens.
 *
 * @param state - The VisitorState object managing the state during the visit.
 * @param node - The Path node representing the path.
 * @param tokens - The array of Identifier or Separator tokens representing the path segments.
 * @param startIndex - The index indicating the start of the range in the tokens array.
 */
function pushMissingPathEscapingDiagnostic(
    state: VisitorState,
    node: Path,
    tokens: (Identifier | Separator)[],
    startIndex: number
): void {
    const start = tokens[startIndex];
    const end = tokens.slice(-1)[0];
    if (!start?.range || !end?.range) {
        return;
    }

    const pathSeparatorDiagnostic: ODataPathSeparatorDiagnostic = {
        rule: ODATA_PATH_SEPARATOR_RULE,
        message: i18n.t('Path_escaping_brackets_should_be_used'),
        range: Range.create(copyPosition(start.range.start), copyPosition(end.range.end)),
        severity: DiagnosticSeverity.Warning
    };

    const valueToBeEscaped = '@' + node.value.split('@').slice(1).join('@');

    const prefix = start.type === IDENTIFIER_TYPE && start.quoted ? '' : '![';
    const suffix = ']';
    // If there are existing escape sequences, we need to remove them
    const body = valueToBeEscaped.replace(/[![\]]/g, '');

    pathSeparatorDiagnostic.data = {
        value: valueToBeEscaped,
        proposedValue: prefix + body + suffix
    };

    state.addDiagnostic(pathSeparatorDiagnostic);
}

/**
 * Pushes a diagnostic for a wrong path separator in the given token range.
 *
 * @param state - The VisitorState object managing the state during the visit.
 * @param token - The Separator token representing the path separator.
 */
function pushWrongPathSeparatorDiagnostic(state: VisitorState, token: Separator): void {
    if (!token.range) {
        return;
    }
    // code action to correct wrong separator
    const message = i18n.t('Wrong_path_separator_0_Did_you_mean_1', {
        currentValue: '/',
        proposedValue: '.',
        interpolation: { escapeValue: false }
    });

    const diagnostic = Diagnostic.create(token.range, message, DiagnosticSeverity.Warning);
    diagnostic.data = {
        caseCheck: {
            value: '/',
            proposedValue: '.',
            lookupPath: [],
            isNamespacedValue: false
        }
    };
    state.addDiagnostic(diagnostic);
}
