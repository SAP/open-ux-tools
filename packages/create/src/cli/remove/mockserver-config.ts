import type { Command } from 'commander';
import prompts from 'prompts';
import { removeMockserverConfig } from '@sap-ux/mockserver-config-writer';
import { getLogger, setLogLevelVerbose, traceChanges } from '../../tracing';
import { hasFileDeletes, validateBasePath } from '../../validation';

/**
 * Add the "add mockserver config" command to a passed command.
 *
 * @param cmd - commander command for adding mockserver config command
 */
export function addRemoveMockserverConfigCommand(cmd: Command): void {
    cmd.command('mockserver-config [path]')
        .option('-v, --verbose', 'show verbose information')
        .option('-f, --force', 'do not ask for confirmation when deleting files')
        .action(async (path, options) => {
            if (options.verbose === true) {
                setLogLevelVerbose();
            }
            await removeMockserverConfiguration(path || process.cwd(), !!options.force);
        });
}

/**
 * Removes a mockserver config from an app or project.
 *
 * @param basePath - path to application root
 * @param force - if true, do not ask before deleting files; otherwise ask
 */
async function removeMockserverConfiguration(basePath: string, force: boolean) {
    const logger = getLogger();
    try {
        logger.debug(`Called remove mockserver-config for path '${basePath}', force is '${force}'`);
        validateBasePath(basePath);
        const fs = removeMockserverConfig(basePath);
        await traceChanges(fs);
        const hasDeletions = hasFileDeletes(fs);
        let doCommit = true;
        if (hasDeletions && !force) {
            doCommit = (
                await prompts([
                    {
                        type: 'confirm',
                        name: 'doCommit',
                        message: `Do you want to apply the changes?`
                    }
                ])
            ).doCommit;
        }
        if (doCommit) {
            fs.commit(() => {
                logger.info(`Mockserver config removed`);
            });
        }
    } catch (error) {
        logger.error(`Error while executing remove mockserver-config '${(error as Error).message}'`);
        logger.debug(error as Error);
    }
}
