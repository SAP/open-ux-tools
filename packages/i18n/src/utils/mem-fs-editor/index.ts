export { readFile } from './read';
export { writeFile } from './write';
