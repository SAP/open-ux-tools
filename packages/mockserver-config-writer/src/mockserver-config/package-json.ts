import { join } from 'path';
import type { Editor } from 'mem-fs-editor';
import type { Package } from '@sap-ux/project-access';
import type { PackageJsonMockConfig } from '../types';

/**
 * Enhance the package.json with dependency for mockserver.
 *
 * @param fs - mem-fs reference to be used for file access
 * @param basePath - path to application root, where package.json is
 * @param config - optional config for mockserver
 */
export function enhancePackageJson(fs: Editor, basePath: string, config?: PackageJsonMockConfig): void {
    const packageJsonPath = join(basePath, 'package.json');
    const packageJson = fs.readJSON(packageJsonPath) as Package;
    enhanceDependencies(packageJson, config?.mockserverModule, config?.mockserverVersion);
    enhanceScripts(fs, packageJson);
    fs.writeJSON(packageJsonPath, packageJson);
}

/**
 * Add mockserver dependencies to package.json content.
 * Takes also care of removing deprecated old module.
 *
 * @param packageJson - parsed package.json content
 * @param mockserverModule - npm name of the mockserver module
 * @param version - npm version string
 */
function enhanceDependencies(
    packageJson: Package,
    mockserverModule = '@sap-ux/ui5-middleware-fe-mockserver',
    version = '2'
) {
    packageJson.devDependencies = packageJson.devDependencies || {};
    delete packageJson.devDependencies['@sap/ux-ui5-fe-mockserver-middleware'];
    packageJson.devDependencies[mockserverModule] = version;
    packageJson.ui5 ||= {};
    packageJson.ui5.dependencies ||= [];
    const ui5Dependencies = packageJson.ui5.dependencies.filter(
        (dep) => dep !== '@sap/ux-ui5-fe-mockserver-middleware'
    );
    if (!ui5Dependencies.includes(mockserverModule)) {
        ui5Dependencies.push(mockserverModule);
    }
    packageJson.ui5.dependencies = ui5Dependencies;
}

/**
 * Add or update start-mock script to package.json.
 *
 * @param fs - mem-fs reference to be used for file access
 * @param packageJson - path to package.json
 */
function enhanceScripts(fs: Editor, packageJson: Package): void {
    packageJson.scripts ||= {};
    packageJson.scripts['start-mock'] =
        copyStartScript(packageJson.scripts.start) || `fiori run --config ./ui5-mock.yaml --open \"/\"`;
}

/**
 * Return a copy of package.json's 'start' script with added or replaced config pointing to ui5-mock.yaml.
 * In case start script can't be copied or is undefined, return undefined.
 *
 * @param startScript - start script from package.json
 * @returns - copy of start script with config to ui5.yaml, undefined if start script can't be copied
 */
function copyStartScript(startScript: string | undefined): string | undefined {
    if (typeof startScript !== 'string') {
        return undefined;
    }
    const fioriRun = 'fiori run';
    const fioriRunIndex = startScript.indexOf(fioriRun);
    if (fioriRunIndex < 0) {
        return undefined;
    }
    const configStartIndex = startScript.indexOf('--config', fioriRunIndex);
    const startMockScript =
        configStartIndex < 0
            ? startScript.replace(fioriRun, `${fioriRun} --config ./ui5-mock.yaml`)
            : replaceConfig(startScript, configStartIndex);
    return startMockScript;
}

/**
 * Replace the --config <any/path> in script with --config ./ui5-mock.yaml and return as new string.
 *
 * @param startScript - script that contains --config path/to/any.yaml
 * @param configStartIndex - index in string where --config starts (first after fiori run)
 * @returns - new script string with replaced --config
 */
function replaceConfig(startScript: string, configStartIndex: number): string {
    const argStart = configStartIndex + '--config'.length + 1;
    const yamlStart = startScript.slice(argStart).search(/[^\s]/) + argStart;
    let separator = ' ';
    let quotischSepOffset = 0;
    if (startScript[yamlStart] === '"' || startScript[yamlStart] === "'") {
        separator = startScript[yamlStart];
        quotischSepOffset = 1;
    }
    let yamlEnd = startScript.indexOf(separator, yamlStart + quotischSepOffset);
    yamlEnd = yamlEnd === -1 ? startScript.length : yamlEnd + quotischSepOffset;
    const startMockScript = `${startScript.substring(0, argStart)}./ui5-mock.yaml${startScript.substring(yamlEnd)}`;
    return startMockScript;
}

/**
 * Remove mockserver script and dependencies from package.json.
 *
 * @param fs - mem-fs reference to be used for file access
 * @param basePath - path to application root, where package.json is
 */
export function removeFromPackageJson(fs: Editor, basePath: string): void {
    const packageJsonPath = join(basePath, 'package.json');
    const packageJson = fs.readJSON(packageJsonPath) as Package;
    delete packageJson.scripts?.['start-mock'];
    if (packageJson.scripts && Object.keys(packageJson.scripts).length === 0) {
        delete packageJson.scripts;
    }
    delete packageJson.devDependencies?.['@sap/ux-ui5-fe-mockserver-middleware'];
    delete packageJson.devDependencies?.['@sap-ux/ui5-middleware-fe-mockserver'];
    if (packageJson.devDependencies && Object.keys(packageJson.devDependencies).length === 0) {
        delete packageJson.devDependencies;
    }
    if (packageJson.ui5?.dependencies && Array.isArray(packageJson.ui5.dependencies)) {
        packageJson.ui5.dependencies = packageJson.ui5.dependencies.filter(
            (d) => d !== '@sap/ux-ui5-fe-mockserver-middleware' && d !== '@sap-ux/ui5-middleware-fe-mockserver'
        );
        if (packageJson.ui5.dependencies.length === 0) {
            delete packageJson.ui5.dependencies;
        }
    }
    if (packageJson.ui5 && Object.keys(packageJson.ui5).length === 0) {
        delete packageJson.ui5;
    }
    fs.writeJSON(packageJsonPath, packageJson);
}
