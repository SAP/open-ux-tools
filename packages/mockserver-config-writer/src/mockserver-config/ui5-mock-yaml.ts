import { join } from 'path';
import type { Editor } from 'mem-fs-editor';
import { UI5Config } from '@sap-ux/ui5-config';
import type { CustomMiddleware } from '@sap-ux/ui5-config';
import type { Manifest } from '@sap-ux/project-access';
import type { Ui5MockYamlConfig } from '../types';
import type { MockserverConfig } from '@sap-ux/ui5-config/dist/types';
import { getMainServiceDataSource } from '../app-info';

/**
 * Enhance or create the ui5-mock.yaml with mockserver config.
 * Following enhancement strategy is applied:
 *
 * ui5-mock.yaml exists
 * -----------------------------
 * Update or create middleware configuration for 'sap-fe-mockserver' with the latest configuration.
 * Other parts of the ui5-mock.yaml will not be modified.
 *
 * ui5-mock.yaml does not exist
 * -----------------------------
 * If ui5.yaml exists, copy it and add/update the 'sap-fe-mockserver' middleware configuration.
 * If ui5.yaml does not exist, create a new ui5-mock.yaml file with 'sap-fe-mockserver' middleware
 * configuration and basic default configuration for 'fiori-tools-proxy' and 'fiori-tools-appreload'.
 *
 * @param fs - mem-fs reference to be used for file access
 * @param basePath - path to project root, where package.json and ui5.yaml is
 * @param webappPath - path to webapp folder, where manifest.json is
 * @param config - optional config passed in by consumer
 */
export async function enhanceYaml(
    fs: Editor,
    basePath: string,
    webappPath: string,
    config?: Ui5MockYamlConfig
): Promise<void> {
    const ui5MockYamlPath = join(basePath, 'ui5-mock.yaml');
    let mockConfig;
    const manifest = fs.readJSON(join(webappPath, 'manifest.json')) as Partial<Manifest> as Manifest;
    const mockserverPath = config?.path || getMainServiceDataSource(manifest)?.uri;

    if (fs.exists(ui5MockYamlPath)) {
        mockConfig = await updateUi5MockYamlConfig(fs, ui5MockYamlPath, mockserverPath);
    } else {
        mockConfig = fs.exists(join(basePath, 'ui5.yaml'))
            ? await generateUi5MockYamlBasedOnUi5Yaml(fs, basePath, mockserverPath)
            : await generateNewUi5MockYamlConfig(manifest['sap.app']?.id || '', mockserverPath);
    }
    const yaml = mockConfig.toString();
    fs.write(ui5MockYamlPath, yaml);
}

/**
 * Update existing ui5-mock.yaml config. This will add or replace existing middleware configuration
 * 'sap-fe-mockserver' with state of the art config.
 *
 * @param fs - Editor instance to read existing information
 * @param ui5MockYamlPath - path to ui5-mock.yaml file
 * @param [path] - optional url path the mockserver listens to
 */
async function updateUi5MockYamlConfig(fs: Editor, ui5MockYamlPath: string, path?: string): Promise<UI5Config> {
    const existingUi5MockYamlConfig = await UI5Config.newInstance(fs.read(ui5MockYamlPath));
    existingUi5MockYamlConfig.updateCustomMiddleware(await getNewMockserverMiddleware(path));
    return existingUi5MockYamlConfig;
}

/**
 * Create a new ui5-mock.yaml based on existing ui5.yaml.
 *
 * @param fs - Editor instance to read existing information
 * @param basePath -
 * @param [path] - optional path for mockserver config
 */
async function generateUi5MockYamlBasedOnUi5Yaml(fs: Editor, basePath: string, path?: string): Promise<UI5Config> {
    const ui5MockYamlConfig = await UI5Config.newInstance(fs.read(join(basePath, 'ui5.yaml')));
    ui5MockYamlConfig.updateCustomMiddleware(await getNewMockserverMiddleware(path));
    return ui5MockYamlConfig;
}

/**
 * Create fresh ui5-mock.yaml configuration which can be stringified and written.
 *
 * @param appId - application id
 * @param [path] - optional url path the mockserver listens to
 */
async function generateNewUi5MockYamlConfig(appId: string, path?: string): Promise<UI5Config> {
    const ui5MockYaml = await UI5Config.newInstance(
        '# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json\n\nspecVersion: "2.5"'
    );
    ui5MockYaml.setMetadata({ name: appId });
    ui5MockYaml.setType('application');
    ui5MockYaml.addFioriToolsProxydMiddleware({ ui5: {} });
    ui5MockYaml.addFioriToolsAppReloadMiddleware();
    ui5MockYaml.addMockServerMiddleware(path);
    return ui5MockYaml;
}

/**
 * Return new mockserver middleware.
 *
 * @param [path] - optional path for mockserver config
 * @returns - mockserver middleware
 */
async function getNewMockserverMiddleware(path?: string): Promise<CustomMiddleware<MockserverConfig>> {
    const ui5MockYaml = await UI5Config.newInstance('');
    ui5MockYaml.addMockServerMiddleware(path);
    const mockserverMiddleware = ui5MockYaml.findCustomMiddleware('sap-fe-mockserver');
    if (!mockserverMiddleware) {
        throw Error('Could not create new mockserver config');
    }
    return mockserverMiddleware as CustomMiddleware<MockserverConfig>;
}

/**
 * Delete ui5-mock.yaml file from mem-fs.
 *
 * @param fs - mem-fs reference to be used for file access
 * @param basePath - path to project root, where package.json and ui5.yaml is
 */
export function removeUi5MockYaml(fs: Editor, basePath: string): void {
    const ui5MockYamlPath = join(basePath, 'ui5-mock.yaml');
    if (fs.exists(ui5MockYamlPath)) {
        fs.delete(ui5MockYamlPath);
    }
}
