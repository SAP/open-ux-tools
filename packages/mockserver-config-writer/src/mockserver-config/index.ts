import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import type { Editor } from 'mem-fs-editor';
import type { MockserverConfig } from '../types';
import { enhancePackageJson, removeFromPackageJson } from './package-json';
import { enhanceYaml, removeUi5MockYaml } from './ui5-mock-yaml';

/**
 *  Add mockserver configuration to a UI5 application.
 *
 * @param basePath - the base path where the package.json and ui5.yaml is
 * @param data - configuration of the mockserver
 * @param fs - the memfs editor instance
 * @returns Promise<Editor> - memfs editor instance with updated files
 */
export async function generateMockserverConfig(basePath: string, data: MockserverConfig, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    enhancePackageJson(fs, basePath, data.packageJsonConfig);
    await enhanceYaml(fs, basePath, data.webappPath, data.ui5MockYamlConfig);
    return fs;
}

/**
 * Remove mockserver configuration.
 *
 * @param basePath - the base path where the package.json and ui5.yaml is
 * @param fs - the memfs editor instance
 * @returns Promise<Editor> - memfs editor instance with updated files
 */
export function removeMockserverConfig(basePath: string, fs?: Editor): Editor {
    if (!fs) {
        fs = create(createStorage());
    }
    removeFromPackageJson(fs, basePath);
    removeUi5MockYaml(fs, basePath);
    return fs;
}
