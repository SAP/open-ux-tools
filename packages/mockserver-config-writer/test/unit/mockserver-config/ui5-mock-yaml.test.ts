import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import type { Editor } from 'mem-fs-editor';
import { join } from 'path';
import { UI5Config } from '@sap-ux/ui5-config';
import type { MockserverConfig } from '@sap-ux/ui5-config/dist/types';
import { enhanceYaml } from '../../../src/mockserver-config/ui5-mock-yaml';

describe('Test enhanceYaml()', () => {
    const basePath = join('/');
    const ui5MockYamlPath = join(basePath, 'ui5-mock.yaml');
    const webappPath = join('/webapp');
    const manifestJsonPath = join(webappPath, 'manifest.json');
    const manifestWithMainService = `{"sap.ui5": { "models": { "": { "dataSource": "ds" } } },"sap.app": { "dataSources": { "ds": { "uri": "ds/uri/" } } }}`;
    const mockManifestJson = `{
        "sap.app": {
            "id": "mockserverv2",
            "dataSources": {
                "mainService": {
                    "uri": "/sap/opu/odata/sap/SEPMRA_PROD_MAN/"
                },
                "SEPMRA_PROD_MAN": {
                    "uri": "/sap/opu/odata/IWFND/CATALOGSERVICE;v=2/Annotations(TechnicalName='SEPMRA_PROD_MAN',Version='0001')/$value/",
                    "type": "ODataAnnotation",
                    "settings": {
                        "localUri": "localService/SEPMRA_PROD_MAN.xml"
                    }
                },
                "annotation": {
                    "type": "ODataAnnotation",
                    "uri": "annotations/annotation.xml",
                    "settings": {
                        "localUri": "annotations/annotation.xml"
                    }
                }
            }
        }
    }`;
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('Create new ui5-mock.yaml with annotations from mock manifest.json', async () => {
        const fs = getFs({ [manifestJsonPath]: mockManifestJson });
        await enhanceYaml(fs, basePath, webappPath, { path: '/path/for/new/config' });

        expect(fs.read(ui5MockYamlPath)).toMatchSnapshot();

        const ui5Config = await UI5Config.newInstance(fs.read(ui5MockYamlPath));
        const mockserverConfig = ui5Config.findCustomMiddleware<MockserverConfig>('sap-fe-mockserver');
        expect(mockserverConfig?.configuration.services?.[0].urlPath).toBe('/path/for/new/config');
        expect(mockserverConfig?.configuration.annotations).toEqual([
            {
                localPath: './webapp/localService/SEPMRA_PROD_MAN.xml',
                urlPath:
                    "/sap/opu/odata/IWFND/CATALOGSERVICE;v=2/Annotations(TechnicalName='SEPMRA_PROD_MAN',Version='0001')/$value/"
            },
            { localPath: './webapp/annotations/annotation.xml', urlPath: 'annotations/annotation.xml' }
        ]);
    });

    test('Update ui5-mock.yaml, path from manifest', async () => {
        const fs = getFsWithUi5MockYaml(manifestWithMainService);
        await enhanceYaml(fs, basePath, webappPath);
        expect(fs.read(ui5MockYamlPath)).toMatchSnapshot();
    });

    test('Update ui5-mock.yaml, path from manifest with annotations', async () => {
        const fs = getFsWithUi5MockYaml(mockManifestJson);
        await enhanceYaml(fs, basePath, webappPath);
        expect(fs.read(ui5MockYamlPath)).toMatchSnapshot();
    });

    test('Update old ui5-mock.yaml with given path', async () => {
        const fs = getFsWithUi5MockYaml('{}');
        await enhanceYaml(fs, basePath, webappPath, { path: 'path/to/service' });
        expect(fs.read(ui5MockYamlPath)).toMatchSnapshot();
    });

    test('Create new ui5-mock.yaml based on ui5.yaml, updated with annotations', async () => {
        const fs = getFsWithUi5Yaml(mockManifestJson);
        await enhanceYaml(fs, basePath, webappPath, { path: 'new/path/to/service' });
        expect(fs.read(ui5MockYamlPath)).toMatchSnapshot();
        // additional check of urlPath, even if snapshot test get lightheartedly updated, the urlPath should remain stable.
        const ui5Config = await UI5Config.newInstance(fs.read(ui5MockYamlPath));
        expect(
            ui5Config.findCustomMiddleware<MockserverConfig>('sap-fe-mockserver')?.configuration.services?.[0].urlPath
        ).toBe('new/path/to/service');
        expect(
            ui5Config.findCustomMiddleware<MockserverConfig>('sap-fe-mockserver')?.configuration.annotations
        ).toEqual([
            {
                localPath: './webapp/localService/SEPMRA_PROD_MAN.xml',
                urlPath:
                    "/sap/opu/odata/IWFND/CATALOGSERVICE;v=2/Annotations(TechnicalName='SEPMRA_PROD_MAN',Version='0001')/$value/"
            },
            { localPath: './webapp/annotations/annotation.xml', urlPath: 'annotations/annotation.xml' }
        ]);
    });

    test('Create new ui5-mock.yaml based on ui5.yaml', async () => {
        const fs = getFsWithUi5Yaml('{}');
        await enhanceYaml(fs, basePath, webappPath, { path: 'new/path/to/service' });
        expect(fs.read(ui5MockYamlPath)).toMatchSnapshot();
        // additional check of urlPath, even if snapshot test get lightheartedly updated, the urlPath should remain stable.
        const ui5Config = await UI5Config.newInstance(fs.read(ui5MockYamlPath));
        expect(
            ui5Config.findCustomMiddleware<MockserverConfig>('sap-fe-mockserver')?.configuration.services?.[0].urlPath
        ).toBe('new/path/to/service');
    });

    test('Create new ui5-mock.yaml without app name in manifest.json', async () => {
        const fs = getFs({ [manifestJsonPath]: '{}' });
        await enhanceYaml(fs, basePath, webappPath, { path: '/path/for/new/config' });
        expect(fs.read(ui5MockYamlPath)).toMatchSnapshot();
        // additional check of urlPath, even if snapshot test get lightheartedly updated, the urlPath should remain stable.
        const ui5Config = await UI5Config.newInstance(fs.read(ui5MockYamlPath));
        expect(
            ui5Config.findCustomMiddleware<MockserverConfig>('sap-fe-mockserver')?.configuration.services?.[0].urlPath
        ).toBe('/path/for/new/config');
    });

    test(`Should throw error in case new added middleware can't be found by name 'sap-fe-mockserver'`, async () => {
        jest.spyOn(UI5Config, 'newInstance').mockResolvedValue({
            addMockServerMiddleware: jest.fn(),
            findCustomMiddleware: () => undefined
        } as unknown as UI5Config);
        const fs = getFsWithUi5MockYaml('{}');
        await expect(enhanceYaml(fs, basePath, webappPath)).rejects.toThrow('mockserver');
    });

    function getFs(files: { [path: string]: string }): Editor {
        const fs = create(createStorage());
        for (const filePath in files) {
            fs.write(filePath, files[filePath]);
        }
        return fs;
    }

    function getFsWithUi5MockYaml(manifestContent: string): Editor {
        return getFs({
            [ui5MockYamlPath]: `specVersion: '2.0'
metadata:
  name: 'app'
type: application
server:
  customMiddleware:
  - name: middleware-before
  - name: sap-fe-mockserver
    beforeMiddleware: fiori-tools-proxy
    configuration:
      service:
        urlBasePath: /some/previous/service/uri
        name: ''
        metadataXmlPath: ./webapp/localService/metadata.xml
        mockdataRootPath: ./webapp/localService/data
        generateMockData: true
  - name: middleware-after`,
            [manifestJsonPath]: manifestContent
        });
    }

    function getFsWithUi5Yaml(manifestContent: string): Editor {
        return getFs({
            [join(
                basePath,
                'ui5.yaml'
            )]: `# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json

specVersion: "2.5"
metadata:
  name: dummy.application
type: application
server:
  customMiddleware:
    - name: first-middleware
      firstProp: firstValue
      nested:
        nestedPropA: nestedValueA # comment for nested value A
        list:
        - listValueA
        - listValueB
    - name: second-middleware`,
            [manifestJsonPath]: manifestContent
        });
    }
});
