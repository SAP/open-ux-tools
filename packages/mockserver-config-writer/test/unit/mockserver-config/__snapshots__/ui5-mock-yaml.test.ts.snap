// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test enhanceYaml() Create new ui5-mock.yaml based on ui5.yaml 1`] = `
"# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json

specVersion: \\"2.5\\"
metadata:
  name: dummy.application
type: application
server:
  customMiddleware:
    - name: first-middleware
      firstProp: firstValue
      nested:
        nestedPropA: nestedValueA # comment for nested value A
        list:
          - listValueA
          - listValueB
    - name: second-middleware
    - name: sap-fe-mockserver
      beforeMiddleware: csp
      configuration:
        mountPath: /
        services:
          - urlPath: new/path/to/service
            metadataPath: ./webapp/localService/metadata.xml
            mockdataPath: ./webapp/localService/data
            generateMockData: true
        annotations: []
"
`;

exports[`Test enhanceYaml() Create new ui5-mock.yaml without app name in manifest.json 1`] = `
"# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json

specVersion: \\"2.5\\"
metadata:
  name: ''
type: application
server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        ui5:
          path:
            - /resources
            - /test-resources
          url: https://ui5.sap.com
    - name: fiori-tools-appreload
      afterMiddleware: compression
      configuration:
        port: 35729
        path: webapp
        delay: 300
    - name: sap-fe-mockserver
      beforeMiddleware: csp
      configuration:
        mountPath: /
        services:
          - urlPath: /path/for/new/config
            metadataPath: ./webapp/localService/metadata.xml
            mockdataPath: ./webapp/localService/data
            generateMockData: true
        annotations: []
"
`;

exports[`Test enhanceYaml() Update old ui5-mock.yaml with given path 1`] = `
"specVersion: '2.0'
metadata:
  name: 'app'
type: application
server:
  customMiddleware:
    - name: middleware-before
    - name: sap-fe-mockserver
      beforeMiddleware: csp
      configuration:
        mountPath: /
        services:
          - urlPath: path/to/service
            metadataPath: ./webapp/localService/metadata.xml
            mockdataPath: ./webapp/localService/data
            generateMockData: true
        annotations: []
    - name: middleware-after
"
`;

exports[`Test enhanceYaml() Update ui5-mock.yaml, path from manifest 1`] = `
"specVersion: '2.0'
metadata:
  name: 'app'
type: application
server:
  customMiddleware:
    - name: middleware-before
    - name: sap-fe-mockserver
      beforeMiddleware: csp
      configuration:
        mountPath: /
        services:
          - urlPath: ds/uri
            metadataPath: ./webapp/localService/metadata.xml
            mockdataPath: ./webapp/localService/data
            generateMockData: true
        annotations: []
    - name: middleware-after
"
`;
